### ISSUE TITLE
StrNcmpUsedAsStrPosInspection is ill advised and should probably be removed/disabled by default

### ISSUE THREAD

**Issue by Riimu** [kalessil/phpinspectionsea#259]
The inspection suggests using `strpos` instead of `strncmp`. While this may be understandable from certain code readability perspective, the description for the inspection reads as follows:

>Strncmp invokes additional comparison strategies, which is not needed in the context. Strpos will do the same job, but without any overhead (it's just searching a string).

The claims of "additional comparison strategies" and "overhead" seem to be rather bogus when actually looking into PHP's source code. In fact, when it comes to short strings, internally PHP actually uses C's `memcmp` for both of these functions. And even when it doesn't, the `strpos` clearly comes out the slower of the two since `strncmp` only needs to compare the strings until a non matching character is found, while `strpos` will have to look for the needle in the entire haystack.

Looking at PHP's source, it becomes quite obvious that the additional functionality offered by `strpos` comes at a greater cost.

From PHP source, the `strncmp` is an internal zend function defined as:

```C
ZEND_FUNCTION(strncmp)
{
	zend_string *s1, *s2;
	zend_long len;

	if (zend_parse_parameters(ZEND_NUM_ARGS(), "SSl", &s1, &s2, &len) == FAILURE) {
		return;
	}

	if (len < 0) {
		zend_error(E_WARNING, "Length must be greater than or equal to 0");
		RETURN_FALSE;
	}

	RETURN_LONG(zend_binary_strncmp(ZSTR_VAL(s1), ZSTR_LEN(s1), ZSTR_VAL(s2), ZSTR_LEN(s2), len));
}
```

Which calls the `zend_binary_strncmp` defined as:

```C
ZEND_API int ZEND_FASTCALL zend_binary_strncmp(const char *s1, size_t len1, const char *s2, size_t len2, size_t length) /* {{{ */
{
	int retval;

	if (s1 == s2) {
		return 0;
	}
	retval = memcmp(s1, s2, MIN(length, MIN(len1, len2)));
	if (!retval) {
		return (int)(MIN(length, len1) - MIN(length, len2));
	} else {
		return retval;
	}
}
```

However, `strpos` is defined as:
```C
PHP_FUNCTION(strpos)
{
	zval *needle;
	zend_string *haystack;
	char *found = NULL;
	char  needle_char[2];
	zend_long  offset = 0;

	ZEND_PARSE_PARAMETERS_START(2, 3)
		Z_PARAM_STR(haystack)
		Z_PARAM_ZVAL(needle)
		Z_PARAM_OPTIONAL
		Z_PARAM_LONG(offset)
	ZEND_PARSE_PARAMETERS_END();

	if (offset < 0) {
		offset += (zend_long)ZSTR_LEN(haystack);
	}
	if (offset < 0 || (size_t)offset > ZSTR_LEN(haystack)) {
		php_error_docref(NULL, E_WARNING, "Offset not contained in string");
		RETURN_FALSE;
	}

	if (Z_TYPE_P(needle) == IS_STRING) {
		if (!Z_STRLEN_P(needle)) {
			php_error_docref(NULL, E_WARNING, "Empty needle");
			RETURN_FALSE;
		}

		found = (char*)php_memnstr(ZSTR_VAL(haystack) + offset,
			                Z_STRVAL_P(needle),
			                Z_STRLEN_P(needle),
			                ZSTR_VAL(haystack) + ZSTR_LEN(haystack));
	} else {
		if (php_needle_char(needle, needle_char) != SUCCESS) {
			RETURN_FALSE;
		}
		needle_char[1] = 0;

		found = (char*)php_memnstr(ZSTR_VAL(haystack) + offset,
							needle_char,
							1,
		                    ZSTR_VAL(haystack) + ZSTR_LEN(haystack));
	}

	if (found) {
		RETURN_LONG(found - ZSTR_VAL(haystack));
	} else {
		RETURN_FALSE;
	}
}
```

Which, in turn calls the `php_memnstr` defined as:

```C
static zend_always_inline const char *
zend_memnstr(const char *haystack, const char *needle, size_t needle_len, const char *end)
{
	const char *p = haystack;
	const char ne = needle[needle_len-1];
	ptrdiff_t off_p;
	size_t off_s;

	if (needle_len == 1) {
		return (const char *)memchr(p, *needle, (end-p));
	}

	off_p = end - haystack;
	off_s = (off_p > 0) ? (size_t)off_p : 0;

	if (needle_len > off_s) {
		return NULL;
	}

	if (EXPECTED(off_s < 1024 || needle_len < 9)) {	/* glibc memchr is faster when needle is too short */
		end -= needle_len;

		while (p <= end) {
			if ((p = (const char *)memchr(p, *needle, (end-p+1))) && ne == p[needle_len-1]) {
				if (!memcmp(needle, p, needle_len-1)) {
					return p;
				}
			}

			if (p == NULL) {
				return NULL;
			}

			p++;
		}

		return NULL;
	} else {
		return zend_memnstr_ex(haystack, needle, needle_len, end);
	}
}
```

And in case both the needle and haystack are long enough, the called `zend_memnstr_ex` is defined as:

```C
ZEND_API const char* ZEND_FASTCALL zend_memnstr_ex(const char *haystack, const char *needle, size_t needle_len, const char *end) /* {{{ */
{
	unsigned int td[256];
	register size_t i;
	register const char *p;

	if (needle_len == 0 || (end - haystack) == 0) {
		return NULL;
	}

	zend_memnstr_ex_pre(td, needle, needle_len, 0);

	p = haystack;
	end -= needle_len;

	while (p <= end) {
		for (i = 0; i < needle_len; i++) {
			if (needle[i] != p[i]) {
				break;
			}
		}
		if (i == needle_len) {
			return p;
		}
		if (UNEXPECTED(p == end)) {
			return NULL;
		}
		p += td[(unsigned char)(p[needle_len])];
	}

	return NULL;
}
```

It does not exactly take much analysis to see which one is considerably more complex. Looking at the actual performance in PHP code, the comparison `strncmp($needle, $haystack, strlen($needle)) === 0` compared to `strpos($haystack, $needle) === 0` is roughly equal at best (when the string matches) but several magnitudes slower at worst (say, a nonmatching string consisting of high number of initial letters).

A simple comparison of a worst case scenario for `strpos`:
```php
<?php

$needle = 'foobarbaz';
$haystack = str_repeat('f', 10000);

$times = 100000;

$timer = microtime(true);
for ($i = 0; $i < $times; $i++) {
    strpos($haystack, $needle) === 0;
}
echo 'strpos:  ' . (microtime(true) - $timer) . "\n";

$timer = microtime(true);
for ($i = 0; $i < $times; $i++) {
    strncmp($needle, $haystack, strlen($needle)) === 0;
}
echo 'strncmp: ' . (microtime(true) - $timer) . "\n";
```

Thus, to me it seems that this inspection is a bit ill advised. While it may left in for code readability reasons, suggesting that using `strpos` for performance reasons seems wrong to me and it would probably be best to at least disable the inspection by default.

**Comment by kalessil** [comment]
That's one of the best Bug-Reports I've ever seen, thank you @Riimu =)

In December 2016 I relocated the inspection from Performance to Control Flow group, but the description was not updated at that time. Thank you for pointing to the outdated docs.

IMO the inspection needs to be enabled for at least reasons:
- the warning honestly references to maintainability;
- if the performance degradation pops up then we have big text fragments stored in memory - and this needs to be investigated;

Obviously more accurate docs would are needed.

**Comment by kalessil** [comment]
@samdark : I don't remember exactly, but wasn't it you who once raised the `strncmp vs strpos` topic?

**Comment by samdark** [comment]
@Riimu is absolutely correct https://3v4l.org/WCE1H

Could be me but we've switched from `strpos()` to `strncmp()` in Yii so if it was me, it was other way around i.e. to prefer `strncmp()` if you don't need position.

**Comment by samdark** [comment]
Found it: https://bitbucket.org/kalessil/phpinspectionsea/issues/172/suggests-strpos-instead-of-strncmp

**Comment by kalessil** [comment]
Right, and we are at 2nd round. I'll just drop it then.

**Comment by dryabov** [comment]
When I did tests (seems it was PHP 5.6), `strpos` is faster for short strings (<100 chars), and `strncmp` is faster for long strings (>1000 chars). The problem was that it's necessary to compare time of `strlen`+`strncmp` vs `strpos`, and overhead for function calling was quite high. Maybe they have improved it in PHP7. Then, it would be nice to have "inversed" inspection, i.e. to found `str[i]pos(...)===0` with recommendation to replace it by `strn[case]cmp+strlen`

**Comment by kalessil** [comment]
Hm, my implementation doesn't check strlen usage. 
@dryabov : how the `strlen+strncmp` looks like? Can you share a code fragment please?

**Comment by kalessil** [comment]
The inspection has been dropped - very solid arguments against.
